modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -1028,4 +1087,357 @@ public Set<ReplicationGroupId> startedGroups() {
                 .map(Entry::getKey)
                 .collect(toSet());
     }
+
+    @TestOnly
+    public boolean isReplicaPrimaryOnly(ReplicationGroupId groupId) {
+        return replicaLifecycle.isReplicaPrimaryOnly(groupId);
+    }
+
+    private static class ReplicaLifecycle {
+        private final IgniteLogger log = Loggers.forClass(ReplicaLifecycle.class);
+
+        final Map<ReplicationGroupId, ReplicaLifecycleContext> replicaContexts = new ConcurrentHashMap<>();
+
+        final Executor replicaStartStopPool;
+
+        final ClockService clockService;
+
+        final PlacementDriver placementDriver;
+
+        final ReplicaManager replicaManager;
+
+        String localNodeId;
+
+        ReplicaLifecycle(Executor replicaStartStopPool, ClockService clockService, PlacementDriver placementDriver,
+                ReplicaManager replicaManager) {
+            this.replicaStartStopPool = replicaStartStopPool;
+            this.clockService = clockService;
+            this.placementDriver = placementDriver;
+            this.replicaManager = replicaManager;
+        }
+
+        void start(String localNodeId) {
+            this.localNodeId = localNodeId;
+            placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, this::onPrimaryElected);
+            placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_EXPIRED, this::onPrimaryExpired);
+        }
+
+        private CompletableFuture<Boolean> onPrimaryElected(PrimaryReplicaEventParameters parameters) {
+            ReplicaLifecycleContext context = getContext(parameters.groupId());
+
+            synchronized (context) {
+                if (localNodeId.equals(parameters.leaseholderId())) {
+                    assert context.replicaState != ReplicaState.STOPPED : "Unexpected primary replica state STOPPED [groupId="
+                            + parameters.groupId() + "].";
+
+                    context.reservedForPrimary = true;
+                } else {
+                    context.reservedForPrimary = false;
+
+                    if (context.replicaState == ReplicaState.PRIMARY_ONLY) {
+                        stopReplica(parameters.groupId(), context, context.deferredStopOperation);
```

Comment:
Why stopping on PRIMARY_REPLICA_EXIRED is not enough?

Reply:
because there can be a scenario:
- lease negotiation begins
- replica is reserved
- replica is excluded from assignments and state is transfered to PRIMARY_ONLY
- negotiation fails and another node is elected as leaseholder, here we get a notification through PRIMARY_ELECTED event. Replica is neither in assignments nor primary nor going to become primary and needs to be stopped